### PR TITLE
[expo-modules][android] propose module-level app or activity customization support

### PR DESCRIPTION
**Pull request by Kudo**
# Why

Some modules require additional setup into MainApplication or MainActivity, e.g. expo-splash-screen or react-native-gesture-handler
Traditionally, we used pre-shaped templates or regular expression in config plugin.
This PR propose a way to make MainApplication and MainActivity extensible.

# How

Introducing `ApplicationLifecycleListener` and `ReactActivityLifecycleListener` for expo-modules, an expo-module package could register these two listeners.
For POC, expo-splash-screen in bare-expo is also migrate to the new model.

## Required changes to templates

MainActivity: `ReactActivityDelegate` -> `ReactActivityDelegateWrapper`

```diff
--- a/apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainActivity.java
+++ b/apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainActivity.java
@@ -10,6 +10,8 @@ import com.facebook.react.ReactActivityDelegate;
 import com.facebook.react.ReactRootView;
 import com.swmansion.gesturehandler.react.RNGestureHandlerEnabledRootView;

+import org.unimodules.adapters.react.ReactActivityDelegateWrapper;
+
 import expo.modules.devlauncher.DevLauncherController;
 import expo.modules.devmenu.react.DevMenuAwareReactActivity;
 import expo.modules.splashscreen.SplashScreenImageResizeMode;
@@ -29,7 +31,8 @@ public class MainActivity extends DevMenuAwareReactActivity {
   @Override
   protected ReactActivityDelegate createReactActivityDelegate() {
     Activity activity = this;
-    ReactActivityDelegate delegate = new ReactActivityDelegate(this, getMainComponentName()) {
+    ReactActivityDelegate delegate = new ReactActivityDelegateWrapper(this,
+      new ReactActivityDelegate(this, getMainComponentName()) {
       @Override
       protected ReactRootView createRootView() {
         return new RNGestureHandlerEnabledRootView(MainActivity.this);
@@ -55,7 +58,7 @@ public class MainActivity extends DevMenuAwareReactActivity {
           pref.edit().putBoolean("isOnboardingFinished", true).apply();
         }
       }
-    };
+    });

     if (MainApplication.USE_DEV_CLIENT) {
       return DevLauncherController.wrapReactActivityDelegate(this, () -> delegate);
```

MainApplication: `ReactNativeHost` -> `ReactNativeHostWrapper` and call helper callbacks at onCreate

```diff
--- a/apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainApplication.java
+++ b/apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainApplication.java
@@ -1,6 +1,7 @@
 package dev.expo.payments;

 import android.app.Application;
+import android.content.res.Configuration;

 import com.facebook.react.PackageList;
 import com.facebook.react.ReactApplication;
@@ -8,14 +9,20 @@ import com.facebook.react.ReactNativeHost;
 import com.facebook.react.ReactPackage;
 import com.facebook.soloader.SoLoader;

-import expo.modules.devlauncher.DevLauncherController;
+import org.unimodules.adapters.react.ApplicationLifecycleDispatcher;
+import org.unimodules.adapters.react.ReactNativeHostWrapper;

 import java.util.List;

+import androidx.annotation.NonNull;
+import expo.modules.devlauncher.DevLauncherController;
+
 public class MainApplication extends Application implements ReactApplication {
   static final boolean USE_DEV_CLIENT = false;

-  private final ReactNativeHost mReactNativeHost = new ReactNativeHost(this) {
+  private final ReactNativeHost mReactNativeHost = new ReactNativeHostWrapper(
+    this,
+    new ReactNativeHost(this) {
     @Override
     public boolean getUseDeveloperSupport() {
       return BuildConfig.DEBUG;
@@ -30,7 +37,7 @@ public class MainApplication extends Application implements ReactApplication {
     protected String getJSMainModuleName() {
       return "index";
     }
-  };
+  });

   @Override
   public ReactNativeHost getReactNativeHost() {
@@ -46,5 +53,12 @@ public class MainApplication extends Application implements ReactApplication {
     if (USE_DEV_CLIENT) {
       DevLauncherController.initialize(this, mReactNativeHost);
     }
+    ApplicationLifecycleDispatcher.onApplicationCreate(this);
+  }
+
+  @Override
+  public void onConfigurationChanged(@NonNull Configuration newConfig) {
+    super.onConfigurationChanged(newConfig);
+    ApplicationLifecycleDispatcher.onConfigurationChanged(this, newConfig);
   }
 }
```

### Performance impact

Leveraging the auto-linking mechanism, a notable change is that we need expo-modules package list before react-native bootstrap. In theory, the change may take longer time to show splash screen.
My measurement from `MainApplication.onCreate` to `SplashScreenViewController.show`
Before: 74.2ms
After 80.0ms

There is no meaningful impact to launch time.
 
# Test Plan

bare-expo android app will show splashscreen even removed the `SplashScreen.show` in MainActivity.

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.io and README.md).
- [ ] This diff will work correctly for `expo build` (eg: updated `@expo/xdl`).
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).
- [ ] Update templates
- [ ] Update @expo/configure-splash-screen

### CODE REVIEW

File: packages/@unimodules/core/android/build.gradle [hunk 1]
```diff
@@ -67,4 +67,5 @@ android {
 
 dependencies {
   implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk7:${safeExtGet('kotlinVersion', '1.4.21')}"
+  implementation 'com.facebook.react:react-native:+'
```
Comment (Kudo) [inline packages/@unimodules/core/android/build.gradle:70]: that is for `ReactNativeHostHandler` which need some react-native classes.
considering we are going to migrate as expo-modules, is it okay to add react-native dependency here?

File: packages/@unimodules/core/android/src/main/java/org/unimodules/core/interfaces/ReactNativeHostHandler.kt [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+   * @return custom bundle asset name, or null if not to override
+   */
+  fun getBundleAssetName(): String? {
+    return null
+  }
+
+  /**
+   * Given chance for JSI modules to register, e.g. for react-native-reanimated
+   */
+  fun onRegisterJSIModules(reactApplicationContext: ReactApplicationContext, jsContext: JavaScriptContextHolder) {
```
Comment (Kudo) [inline packages/@unimodules/core/android/src/main/java/org/unimodules/core/interfaces/ReactNativeHostHandler.kt:30]: that is specific for something like react-native-reanimated.
my proposal is to add the `createReactNativeHostHandlers` integration directly in react-native-reanimated.
however, this requires expo-modules, but expo-modules auto-linking  is based on react-native-cli auto-linking.

the question is: if a module both support cli auto-linking and expo-modules auto-linking, how could we deal with this case?

**Review by bbarthec — COMMENTED** [review]
The idea is very neat! 💪
I only need to be sure we're having a comprehensive solution that would not require massive refactors (and by `massive` I mean even 1 line replacement 😅).
I need @tsapeta to look into this and I want to discuss whether we can deliver a holistic solution for hooking into native things (state/events/lifecycle/etc).

File: packages/@unimodules/react-native-adapter/android/src/main/java/org/unimodules/adapters/react/ReactActivityDelegateWrapper.kt [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import org.unimodules.core.interfaces.ReactActivityLifecycleListener
+
+open class ReactActivityDelegateWrapper(activity: ReactActivity, mainComponentName: String?)
+  : ReactActivityDelegate(activity, mainComponentName) {
+  private val reactActivityLifecycleListeners: ArrayList<ReactActivityLifecycleListener> = ArrayList()
+  init {
+    for (pkg in ExpoModulesPackageList.getPackageList()) {
+      reactActivityLifecycleListeners.addAll(pkg.createReactActivityLifecycleListeners(activity))
+    }
+  }
```
[the remark this answers is not in the document]
Comment (Kudo) [reply packages/@unimodules/react-native-adapter/android/src/main/java/org/unimodules/adapters/react/ReactActivityDelegateWrapper.kt:17]: the functional style is neat. i like to pretty much.
the working version is like this, will update the code accordingly.
```kotlin
  private val reactActivityLifecycleListeners: List<ReactActivityLifecycleListener> =
    ExpoModulesPackageList.getPackageList()
      .flatMap { it.createReactActivityLifecycleListeners(activity) }
```
doesn't use `::` yet, but still clear!?

**Comment by Kudo** [comment]
> The idea is very neat! 💪
> I only need to be sure we're having a comprehensive solution that would not require massive refactors (and by `massive` I mean even 1 line replacement 😅).
> I need @tsapeta to look into this and I want to discuss whether we can deliver a holistic solution for hooking into native things (state/events/lifecycle/etc).

thanks. that is still a proposal and welcome for comments.
hopefully we could come out a solid design for modularization.

in the meantime, i will try to migrate other module based on this architecture.
this is [for expo-updates](https://github.com/expo/expo/commit/9c5c2842c5f041a98063bed2b3b2ddcd72b6203c)
typically, we use app's `BuildConfig.DEBUG` to determine if to enable expo-updates.
that is not available for android library, so to use `useDeveloperSupport` instead.

File: apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainActivity.java [hunk 4]
```diff
@@ -29,7 +31,7 @@ protected String getMainComponentName() {
   @Override
   protected ReactActivityDelegate createReactActivityDelegate() {
     Activity activity = this;
-    ReactActivityDelegate delegate = new ReactActivityDelegate(this, getMainComponentName()) {
+    ReactActivityDelegate delegate = new ReactActivityDelegateWrapper(this, getMainComponentName()) {
```
[the remark this answers is not in the document]
Comment (Kudo) [reply apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainActivity.java:34]: share my experiment today:
it is possible for androidLibraryA to add proguard rules for androidLibraryB by `consumerProguardFiles`.
we could keep ReactActivityDelegate class without name mangling and make reflection feasible.
the risk for not mangling `ReactActivityDelegate` or `ReactNativeHost` is low.
if we don't have concern for this approach, i will change the code to use wrapping pattern and reflection accordingly.

File: packages/@unimodules/core/android/src/main/java/org/unimodules/core/interfaces/ReactNativeHostHandler.kt (hunk 2 above)
Comment (Kudo) [reply packages/@unimodules/core/android/src/main/java/org/unimodules/core/interfaces/ReactNativeHostHandler.kt:30]: share yet another experiment today:
gradle will not complain if a dependency be added twice, so it is okay for a module to be both react-native-cli auto-linking and expo-module auto-linking.

for react-native-reanimated and react-native-gesture-handler, we could have an expo-module adapter package only for   application listener hook but no native modules nor view managers.
something like this:

```java
public class ExpoReanimatedAdapterPackage extends BasePackage {
  @Override
  public List<? extends ReactNativeHostHandler> createReactNativeHostHandlers(Context context) {
    return Collections.singletonList(new ReactNativeHostHandler() {
      @Override
      public void onRegisterJSIModules(ReactApplicationContext reactApplicationContext, JavaScriptContextHolder jsContext) {
        new ReanimatedJSIModulePackage().getJSIModules(reactApplicationContext, reactApplicationContext);
      }
    });
  }
}
```

will follow this strategy if there is no concerns.

File: packages/@unimodules/react-native-adapter/android/src/main/java/org/unimodules/adapters/react/ApplicationLifecycleDispatcher.kt [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+import android.content.res.Configuration
+
+class ApplicationLifecycleDispatcher private constructor() {
+  companion object {
+
+    @JvmStatic
+    fun onApplicationCreate(application: Application) {
+      ExpoModulesPackageList.getPackageList()
+        .flatMap { it.createApplicationLifecycleListeners(application) }
+        .forEach { it.onCreate(application) }
```
Comment (lukmccall) [inline packages/@unimodules/react-native-adapter/android/src/main/java/org/unimodules/adapters/react/ApplicationLifecycleDispatcher.kt:13]: For this method you should be able to use a method reference:
```kt
ExpoModulesPackageList.getPackageList()
        .flatMap(Package::createApplicationLifecycleListeners)
        .forEach(ApplicationLifecycleListeners::onCreate)
```
But I don't have a strong option here. So you can leave it as it is.

**Review by lukmccall — COMMENTED** [review]
I think if we want to add a similar mechanism to the app delegates on Android, this is the best way to achieve it. However, I need to think about this a little bit more 😅

Some comments for now:
- we should add `CallSuper`(https://developer.android.com/reference/androidx/annotation/CallSuper) annotation to every method in our wrapper to inform users that they need to call the implementation from super class too (Someone could override those methods and didn't call our implementation). 
- we should think of a situation, where users want to provide their custom bridge initialization.

File: apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainActivity.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (bbarthec) [reply apps/bare-expo/android/app/src/main/java/dev/expo/payments/MainActivity.java]: cc @tsapeta, @lukmccall as we're having a new idea here ⬆️.
My opinion is: let's give it a try 👍

File: packages/@unimodules/react-native-adapter/android/src/main/java/org/unimodules/adapters/react/ReactNativeHostWrapper.kt [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+import com.facebook.react.bridge.ReactMarker
+import com.facebook.react.bridge.ReactMarkerConstants
+import com.facebook.react.common.LifecycleState
+import com.facebook.react.devsupport.RedBoxHandler
+import com.facebook.react.uimanager.UIImplementationProvider
+import java.lang.reflect.Method
+
+class ReactNativeHostWrapper(application: Application, host: ReactNativeHost)
+  : ReactNativeHost(application) {
+  private val host = host
```
Comment (bbarthec) [inline packages/@unimodules/react-native-adapter/android/src/main/java/org/unimodules/adapters/react/ReactNativeHostWrapper.kt:24]:
```suggestion
class ReactNativeHostWrapper(
  application: Application,
  private val host: ReactNativeHost
) : ReactNativeHost(application) {
```
You can inline property in the constructor.

File: packages/@unimodules/react-native-adapter/android/src/main/java/org/unimodules/adapters/react/ReactNativeHostWrapper.kt [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+    return invokeDelegateMethod("getPackages")
+  }
+
+  //endregion
+
+  //region Internals
+
+  inner class JSIModuleContainerPackage : JSIModulePackage {
+    override fun getJSIModules(reactApplicationContext: ReactApplicationContext,
+                               jsContext: JavaScriptContextHolder): List<JSIModuleSpec<JSIModule>> {
```
Comment (bbarthec) [inline packages/@unimodules/react-native-adapter/android/src/main/java/org/unimodules/adapters/react/ReactNativeHostWrapper.kt:121]:
```suggestion
    override fun getJSIModules(
      reactApplicationContext: ReactApplicationContext,
      jsContext: JavaScriptContextHolder
    ): List<JSIModuleSpec<JSIModule>> {
```

File: packages/@unimodules/react-native-adapter/android/src/main/java/org/unimodules/adapters/react/ReactActivityDelegateWrapper.kt [hunk 8]
```diff
@@ -0,0 +1,150 @@
...
+import com.facebook.react.ReactInstanceManager
+import com.facebook.react.ReactNativeHost
+import com.facebook.react.ReactRootView
+import com.facebook.react.modules.core.PermissionListener
+import java.lang.reflect.Method
+
+class ReactActivityDelegateWrapper(activity: ReactActivity, delegate: ReactActivityDelegate)
+  : ReactActivityDelegate(activity, null) {
+  private val delegate = delegate
+  private val activity = activity
```
Comment (bbarthec) [inline packages/@unimodules/react-native-adapter/android/src/main/java/org/unimodules/adapters/react/ReactActivityDelegateWrapper.kt:20]:
```suggestion
class ReactActivityDelegateWrapper(
  private val activity: ReactActivity,
  private val delegate: ReactActivityDelegate
) : ReactActivityDelegate(activity, null) {
```

**Review by bbarthec — COMMENTED** [review]
It looks very promising 👍

File: packages/@unimodules/react-native-adapter/android/src/main/java/org/unimodules/adapters/react/ReactNativeHostWrapper.kt [hunk 9]
```diff
@@ -0,0 +1,105 @@
...
+  private val applicationLifecycleListeners: ArrayList<ApplicationLifecycleListener> = ArrayList()
+  private val reactNativeHostHandlers: ArrayList<ReactNativeHostHandler> = ArrayList()
+  init {
+    for (pkg in ExpoModulesPackageList.getPackageList()) {
+      applicationLifecycleListeners.addAll(pkg.createApplicationLifecycleListeners(application))
+      reactNativeHostHandlers.addAll(pkg.createReactNativeHostHandlers(application))
+    }
+  }
+
+  override fun createReactInstanceManager(): ReactInstanceManager {
```
[the remark this answers is not in the document]
Comment (Kudo) [reply packages/@unimodules/react-native-adapter/android/src/main/java/org/unimodules/adapters/react/ReactNativeHostWrapper.kt:31]: the new method in https://github.com/expo/expo/pull/13457/commits/14115f40c6088090cd1bc16c00f9d6071b0f29ae to handle this case:
```kotlin
  /**
   * Given chance for modules to customize {@link ReactInstanceManager}
   *
   * @return instance of {@link ReactInstanceManager}, or null if not to override
   */
 fun createReactInstanceManager(): ReactInstanceManager? {
    return null
  }
```

**Comment by Kudo** [comment]
> I think if we want to add a similar mechanism to the app delegates on Android, this is the best way to achieve it. However, I need to think about this a little bit more 😅
> 
> Some comments for now:
> 
> * we should add `CallSuper`(https://developer.android.com/reference/androidx/annotation/CallSuper) annotation to every method in our wrapper to inform users that they need to call the implementation from super class too (Someone could override those methods and didn't call our implementation).

an advantage of true wrapper pattern (the latest code with reflection) is that our wrappers to be final classes. methods are not overridable.
i.e.
user could only override ReactNativeHost methods but not ReactNativeWrapperHost methods
user could only override ReactActivityDelegate methods but not ReactActivityDelegateWrapper methods

even user forget callsuper, it is okay b/c our wrapper method would be the first one to be called. 

> * we should think of a situation, where users want to provide their custom bridge initialization.

https://github.com/expo/expo/pull/13457/commits/14115f40c6088090cd1bc16c00f9d6071b0f29ae add a way for this case.

thanks for your great input.
welcome for comments so that we could come out a feasible & extensible design for expo-modules.

File: packages/@unimodules/core/android/src/main/java/org/unimodules/core/interfaces/ReactNativeHostHandler.kt (hunk 2 above)
Comment (EvanBacon) [reply packages/@unimodules/core/android/src/main/java/org/unimodules/core/interfaces/ReactNativeHostHandler.kt:40]: > but expo-modules auto-linking is based on react-native-cli auto-linking

Other way around https://github.com/react-native-community/cli/issues/248#issue-422591744

> will follow this strategy if there is no concerns.

I don't see any immediate issues with this approach, looks fine.

File: packages/expo-splash-screen/android/src/main/java/expo/modules/splashscreen/SplashScreenReactActivityLifecycleListener.kt [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+import org.unimodules.core.interfaces.ReactActivityLifecycleListener
+
+class SplashScreenReactActivityLifecycleListener(activityContext: Context) : ReactActivityLifecycleListener {
+  override fun onCreate(activity: Activity, savedInstanceState: Bundle?) {
+    // To support backward compatible or SplashScreenImageResizeMode customization
+    // that calling `SplashScreen.show()` in MainActivity,
+    // we postpone the in-module call to the end of main loop.
+    // If MainActivity.onCreate has `SplashScreen.show()`, it will override the call here.
+    Handler(activity.mainLooper).post {
+      SplashScreen.show(activity, SplashScreenImageResizeMode.CONTAIN, ReactRootView::class.java, false)
```
Comment (EvanBacon) [inline packages/expo-splash-screen/android/src/main/java/expo/modules/splashscreen/SplashScreenReactActivityLifecycleListener.kt:18]: As specified earlier, maybe we should read these values from the AndroidManifest.xml or other static resource file in the `android/` folder.

Comment (Kudo) [reply packages/expo-splash-screen/android/src/main/java/expo/modules/splashscreen/SplashScreenReactActivityLifecycleListener.kt:18]: yes, plan to have this modularizing expo-splash-screen in another pr.

File: packages/@unimodules/react-native-adapter/android/src/main/java/org/unimodules/adapters/react/ApplicationLifecycleDispatcher.kt (hunk 5 above)
Comment (Kudo) [reply packages/@unimodules/react-native-adapter/android/src/main/java/org/unimodules/adapters/react/ApplicationLifecycleDispatcher.kt:13]: looks like not feasible b/c additional `application` parameter for createApplicationLifecycleListeners and onCreate.
have a playground for this: https://pl.kotl.in/yDauMooMP
does kotlin support partial parameter bound call like javascript bind or c++ std::bind?

Comment (lukmccall) [reply packages/@unimodules/react-native-adapter/android/src/main/java/org/unimodules/adapters/react/ApplicationLifecycleDispatcher.kt:13]: It won't work in that case. Sorry, my bad.

Comment (Kudo) [reply packages/@unimodules/react-native-adapter/android/src/main/java/org/unimodules/adapters/react/ApplicationLifecycleDispatcher.kt:13]: not at all, thank you folks to teach me kotlin best practice from review 🥰

**Review by tsapeta — APPROVED** [review]

**Review by bbarthec — APPROVED** [review]

**Review by brentvatne — APPROVED** [review]

**Comment by Kudo** [comment]
thanks for all the reviews and approvals. i am going to merge this.
will get a chance to migrate expo-splash-screen this model and update expo-template-bare-minimum template together.

**Pull request merged**